internal/configuration/storage/registry/registry.go

```diff
@@ -0,0 +1,167 @@
+
+// Storage is a config storage in a OCI Registry.
+//
+// Configuration ID is a sha256 of the contents, so it matches registry content-addressable storage.
+type Storage struct {
+	pusher     *remote.Pusher
+	puller     *remote.Puller
+	repository name.Repository
+}
+
+// Check interface.
+var _ storage.Storage = (*Storage)(nil)
+
+// digestPrefix is "sha256:".
+var digestPrefix = digest.Canonical.String() + ":"
+
+// NewStorage creates a new storage.
+func NewStorage(repository name.Repository, remoteOpts []remote.Option) (*Storage, error) {
+	s := &Storage{
+		repository: repository,
+	}
+
+	var err error
+
+	s.pusher, err = remote.NewPusher(append(remoteOpts, remote.WithNondistributable)...)
+	if err != nil {
+		return nil, err
+	}
+
+	s.puller, err = remote.NewPuller(remoteOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+// Head checks if the configuration exists.
+func (s *Storage) Head(ctx context.Context, id string) error {
+	// pre-validate the ID, so that invalid IDs return not found
+	_, err := v1.NewHash(digestPrefix + id)
+	if err != nil {
+		return xerrors.NewTaggedf[storage.ErrNotFoundTag]("configuration ID %q not found", id)
+	}
+
+	layer, err := s.puller.Layer(ctx, s.repository.Digest(digestPrefix+id))
+	if err != nil {
+		return err
+	}
+
+	// check if the layer exists
+	// we're not interested in size, but it calls HEAD on the layer
+	_, err = layer.Size()
+	if err == nil {
+		return nil
+	}
+
+	var transportError *transport.Error
+
+	if errors.As(err, &transportError) && transportError.StatusCode == http.StatusNotFound {
+		return xerrors.NewTaggedf[storage.ErrNotFoundTag]("configuration ID %q not found", id)
+	}
+
+	return err
+}
+
+// Get returns the configuration.
+func (s *Storage) Get(ctx context.Context, id string) ([]byte, error) {
+	// pre-validate the ID
+	_, err := v1.NewHash(digestPrefix + id)
+	if err != nil {
+		return nil, xerrors.NewTaggedf[storage.ErrNotFoundTag]("configuration ID %q not found", id)
+	}
+
+	layer, err := s.puller.Layer(ctx, s.repository.Digest(digestPrefix+id))
+	if err != nil {
+		var transportError *transport.Error
+
+		if errors.As(err, &transportError) && transportError.StatusCode == http.StatusNotFound {
+			return nil, xerrors.NewTaggedf[storage.ErrNotFoundTag]("configuration ID %q not found", id)
+		}
+
+		return nil, err
+	}
+
+	// pull the layer
+	r, err := layer.Compressed()
+	if err != nil {
+		return nil, err
+	}
+
+	defer r.Close() //nolint:errcheck
+
+	return io.ReadAll(r)
+}
+
+// layerWrapper adapts to the expected v1.Layer interface.
+type layerWrapper struct {
+	id   string
+	data []byte
+}
+
+// Digest returns the Hash of the compressed layer.
+func (w *layerWrapper) Digest() (v1.Hash, error) {
+	return v1.Hash{
+		Algorithm: digest.Canonical.String(),
+		Hex:       w.id,
+	}, nil
+}
+
+// Compressed returns an io.ReadCloser for the compressed layer contents.
+//
+// In fact, it returns raw contents.
```

Comment:
is this because we just upload the raw config?

Reply:
it's the interface of the library we use to access the registry, it's a bit more specific to "real" images, but we can still store raw data using it